File: OC/内容层动画/WGOCCoreAnimation/WGOCCoreAnimation/水纹动画/WGWaterWaveBtn.m
//
//  WGWaterWaveBtn.m
//  WGOCCoreAnimation
//
//  Created by wanggang on 2019/5/22.
//  Copyright © 2019 bozhong. All rights reserved.
//

#import "WGWaterWaveBtn.h"

@interface WGWaterWaveBtn()

@property (nonatomic, strong) NSTimer *timer;

@end

@implementation WGWaterWaveBtn


- (void)startAniWithEvent:(UIEvent *)event{
    
    self.userInteractionEnabled = NO;
    NSSet *touchSet = [event touchesForView:self];
    NSArray *touchArr = touchSet.allObjects;
    UITouch *touch = touchArr.firstObject;
    CGPoint point = [touch locationInView:self];
    self.touchPoint = point;
    
    self.timer = [NSTimer timerWithTimeInterval:0.02 target:self selector:@selector(timerMethod) userInfo:nil repeats:YES];
    [[NSRunLoop currentRunLoop] addTimer:_timer forMode:NSDefaultRunLoopMode];
}

- (void)timerMethod{
    
    self.viewRadius += 5;
    [self setNeedsDisplay];
    
    if (self.viewRadius >= self.width) {
        [self.timer invalidate];
        self.timer = nil;
        self.viewRadius = 0;
        [self setNeedsDisplay];
        self.userInteractionEnabled = YES;
    }
}

- (void)drawRect:(CGRect)rect {
    
    CGContextRef context = UIGraphicsGetCurrentContext();
    CGContextAddArc(context, self.touchPoint.x, self.touchPoint.y, self.viewRadius, 0, M_PI*2, NO);
    [self.aniColor setStroke];
    [self.aniColor setFill];
    CGContextFillPath(context);
}

-(UIColor *)aniColor{
    if (!_aniColor) {
        _aniColor = [UIColor cyanColor];
    }
    return _aniColor;
}

@end

File: Swift/内容层动画/WGCoreAnimation/WGCoreAnimation/水纹动画/WGWaterWaveBtn.swift
//
//  WGWaterWaveBtn.swift
//  WGCoreAnimation
//
//  Created by wanggang on 2019/4/23.
//  Copyright © 2019 wanggang. All rights reserved.
//

import UIKit

class WGWaterWaveBtn: UIButton {

    //水纹半径
    var viewRadius:CGFloat = 0.0
    //定时器
    var timer: Timer?
    //点击处
    var pointX:CGFloat = 0.0
    var pointY:CGFloat = 0.0
    //水纹颜色
    var aniColor = UIColor.cyan
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        backgroundColor = UIColor.brown
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func startAnimation(event: UIEvent) {
        
        self.isUserInteractionEnabled = false
        let buttonView = self as UIView
        let touchSet = event.touches(for: buttonView)! as NSSet
        let touchArr = touchSet.allObjects
        let touch = touchArr.first as! UITouch
        let touchPoint = touch.location(in: buttonView)
        
        pointX = touchPoint.x
        pointY = touchPoint.y
        
        timer = Timer(timeInterval: 0.02, target: self, selector: #selector(timerMethod), userInfo: nil, repeats: true)
        RunLoop.main.add(timer!, forMode: .common)
    }
    
    @objc func timerMethod() {
        
        self.viewRadius += 5
        self.setNeedsDisplay()
        
        if viewRadius >= self.width {
            timer?.invalidate()
            viewRadius = 0
            self.setNeedsDisplay()
            self.isUserInteractionEnabled = true
        }
    }
    
    override func draw(_ rect: CGRect) {
        
        let ctx = UIGraphicsGetCurrentContext()
        ctx?.addArc(center: CGPoint(x: pointX, y: pointY), radius: viewRadius, startAngle: 0, endAngle: .pi*2, clockwise: false)
        aniColor.setStroke()
        aniColor.setFill()
        ctx?.fillPath()
    }

}
